src/main/scala/chisel3/tester/TreadleBackend.scala

```diff
@@ -237,68 +245,48 @@ object TreadleExecutive {
-      case _ =>
-    }
-    optionsManager.firrtlOptions = optionsManager.firrtlOptions.copy(compilerName = "low")
+    // Force a cleanup: long SBT runs tend to fail with memory issues
+    System.gc()
 
-    execOptions foreach {
-      case userOptions: HasTreadleSuite => optionsManager.treadleOptions = userOptions.treadleOptions
-      case _ =>
-    }
+    try {
+      val generatorAnnotation = chisel3.stage.ChiselGeneratorAnnotation(dutGen)
 
-    // Tester options take priority over exec options
-    val testName = testOptions.name.replaceAll(" ", "_").replaceAll("\\W+", "")  // sanitize filename
-    optionsManager.commonOptions = optionsManager.commonOptions.copy(
-        targetDirName = s"test_run_dir/$testName")
-    if (testOptions.writeVcd) {
-      optionsManager.treadleOptions = optionsManager.treadleOptions.copy(writeVCD = true)
-    }
+      val chiseledAnnotations = (new ChiselStage).run(
+        annotationSeq ++ Seq(generatorAnnotation, NoRunFirrtlCompilerAnnotation)
+//        annotationSeq ++ Seq(generatorAnnotation)
+      )
 
-    // Force a cleanup: long SBT runs tend to fail with memory issues
-    System.gc()
+      val circuit = generatorAnnotation.elaborate.circuit
 
-    chisel3.Driver.execute(optionsManager, dutGen) match {
-      case ChiselExecutionSuccess(Some(circuit), _, Some(firrtlExecutionResult)) =>
-        firrtlExecutionResult match {
-          case success: FirrtlExecutionSuccess =>
-            val dut = getTopModule(circuit).asInstanceOf[T]
-            optionsManager.firrtlOptions = optionsManager.firrtlOptions.copy(
-              annotations = success.circuitState.annotations.toSeq.toList
-            )
-            val interpretiveTester = new TreadleTester(success.emitted, optionsManager)
-
-            val portNames = DataMirror.modulePorts(dut).flatMap { case (name, data) =>
-              getDataNames(name, data).toList
-            }.toMap
-            val paths = success.circuitState.annotations.collect {
-              case c: CombinationalPath => c
-            }
-            val pathsAsData = combinationalPathsToData(dut, paths, portNames)
-
-            new TreadleBackend(dut, portNames, pathsAsData, interpretiveTester)
-          case FirrtlExecutionFailure(message) =>
-            throw new Exception(s"FirrtlBackend: failed firrtl compile message: $message")
-        }
-      case _ =>
-        throw new Exception("Problem with compilation")
+      val dut = getTopModule(circuit).asInstanceOf[T]
+
+      val portNames = DataMirror.modulePorts(dut).flatMap { case (name, data) =>
+        getDataNames(name, data).toList
+      }.toMap
+
+      val treadledAnnotations = TreadleTesterPhase.transform(chiseledAnnotations)
+
+      val treadleTester = treadledAnnotations.collectFirst { case TreadleTesterAnnotation(t) => t }.getOrElse(
+        throw new Exception(s"TreadleTesterPhase did not build a treadle tester")
```

Comment:
Any reason this handles the get failure with a custom Exception while the next one (circuitState) just does a get?

Reply:
The error handling could be improved. In treadle there are a number of ways that it can fail to produce a tester, but if it does produce a tester there should not be a case where a circuit state was not generated. I have made the message more explicit in that it will now show the annotations involved. This will help me, at least, debug this mostly unlikely case.

Reply:
Unless it's common enough that you think a user might encounter it frequently, I'm fine with it dumping a stack trace, which will have enough information for someone to open an issue.

Reply:
Also, would it make sense to just have the TreadleTesterPhase throw an exception? If there's a fatal error in there, it seems like it would make sense to terminate there and provide a stack trace from inside treadle?

Reply:
I have changed this to show the annotations that were sent in. I am not entirely sure what roles treadle phase will take on. So I am hesitant to make it an exception internally. I think this will make debugging easiest.

Reply:
My understanding of this is that if you give treadle an empty annotation seq, it will do nothing to it. Hence, this PR needs to check if an expected output annotation is present. This contrasts with FIRRTL, where there is an explicit `Checks` phase that validates the input annotation seq and will, e.g., throw an exception internally if the input annotation seq doesn't make sense. You may consider that pattern for treadle with a checks phase.

Reply:
I think it's also useful to look at it from a user intent point of view: what do you expect the user to expect when TreadleTesterPhase is called? If it's expected to do one thing, and it can't do that one thing, then that should probably be a hard error. If the invocation is super generalized and might do totally different things (including nothing - by design and "contract"), then this strategy probably won't work, but you might also want to rethink the interface there.

Reply:
I'd like to leave it the way it is for the moment. TreadlePhase is used by the TreadleRepl which can be spun up without specifying firrtl source. Old testers needs to be converted as well and it has it has its own pattern, perhaps that will help drive where to put the error and what it should be.